src/librustc/ty/layout.rs

```diff
@@ -872,56 +890,118 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
+                    let mut max_size = Size::ZERO;
+                    let mut second_max_size = Size::ZERO;
+
+                    // The size computations below assume that the padding is minimum.
+                    // This is the case when fields are re-ordered.
+                    let struct_reordering_opt = !def.repr.inhibit_struct_field_reordering_opt();
+                    if struct_reordering_opt {
+                        let mut extend_niche_range = |d| {
+                            niche_variants =
+                                *niche_variants.start().min(&d)..=*niche_variants.end().max(&d);
+                        };
+                        let mut align = dl.aggregate_align;
 
-                    // Find one non-ZST variant.
-                    'variants: for (v, fields) in variants.iter_enumerated() {
-                        if absent(fields) {
-                            continue 'variants;
+                        // Find the largest and second largest variant.
+                        for (v, fields) in variants.iter_enumerated() {
+                            if absent(fields) {
+                                continue;
+                            }
+                            let mut size = Size::ZERO;
+                            for &f in fields {
+                                align = align.max(f.align);
+                                size += f.size;
+                            }
+                            if size > max_size {
+                                second_max_size = max_size;
+                                max_size = size;
+                                if let Some(d) = dataful_variant {
+                                    extend_niche_range(d);
+                                }
+                                dataful_variant = Some(v);
+                            } else if size == max_size {
+                                if let Some(d) = dataful_variant {
+                                    extend_niche_range(d);
+                                }
+                                dataful_variant = None;
+                                extend_niche_range(v);
+                            } else {
+                                second_max_size = second_max_size.max(size);
+                                extend_niche_range(v);
+                            }
                         }
-                        for f in fields {
-                            if !f.is_zst() {
-                                if dataful_variant.is_none() {
-                                    dataful_variant = Some(v);
-                                    continue 'variants;
-                                } else {
-                                    dataful_variant = None;
-                                    break 'variants;
+                        if !max_size.is_aligned(align.abi) {
+                            // Don't perform niche optimisation if there is padding anyway
```

Comment:
I believe this fixes  #63866.   If it detects that the variant would contains padding, the padding will be used for the discriminant.

Reply:
Maybe, but it seems harder to reason about than having two candidates for a layout and picking the best one out of them.

Reply:
I'm thinking computing the second candidate would be a waste of time.